File: server/src/collectors/classifieds.js
/**
 * Japanese Classifieds Collector
 * Aggregates listings from Japan's equivalent of Craigslist:
 * - Jmty (ジモティー) - local classifieds, free items, barter
 * - Mercari - marketplace listings
 * - Yahoo! Auctions Japan - auction items
 * Scrapes public listings with geo data
 */

const LISTING_CATEGORIES = ['furniture', 'electronics', 'clothing', 'vehicles', 'free_items', 'barter', 'pets', 'tickets', 'sports', 'books', 'musical_instruments', 'handmade'];
const PLATFORMS = ['jmty', 'mercari', 'yahoo_auction'];

const JAPAN_AREAS = [
  // Tokyo 23 wards
  { area: '世田谷区', pref: '東京都', lat: 35.6461, lon: 139.6531, pop: 920000 },
  { area: '練馬区', pref: '東京都', lat: 35.7356, lon: 139.6518, pop: 740000 },
  { area: '大田区', pref: '東京都', lat: 35.5613, lon: 139.7161, pop: 730000 },
  { area: '足立区', pref: '東京都', lat: 35.7752, lon: 139.8046, pop: 690000 },
  { area: '江戸川区', pref: '東京都', lat: 35.6928, lon: 139.8682, pop: 700000 },
  { area: '杉並区', pref: '東京都', lat: 35.6994, lon: 139.6365, pop: 580000 },
  { area: '板橋区', pref: '東京都', lat: 35.7512, lon: 139.7090, pop: 580000 },
  { area: '新宿区', pref: '東京都', lat: 35.6938, lon: 139.7036, pop: 350000 },
  { area: '渋谷区', pref: '東京都', lat: 35.6640, lon: 139.6982, pop: 230000 },
  { area: '中野区', pref: '東京都', lat: 35.7073, lon: 139.6638, pop: 340000 },
  { area: '豊島区', pref: '東京都', lat: 35.7263, lon: 139.7171, pop: 290000 },
  { area: '目黒区', pref: '東京都', lat: 35.6338, lon: 139.6980, pop: 280000 },
  // Kanagawa
  { area: '横浜市', pref: '神奈川県', lat: 35.4437, lon: 139.6380, pop: 3750000 },
  { area: '川崎市', pref: '神奈川県', lat: 35.5309, lon: 139.7030, pop: 1540000 },
  { area: '相模原市', pref: '神奈川県', lat: 35.5710, lon: 139.3731, pop: 720000 },
  // Osaka
  { area: '大阪市北区', pref: '大阪府', lat: 34.7055, lon: 135.4983, pop: 130000 },
  { area: '大阪市中央区', pref: '大阪府', lat: 34.6813, lon: 135.5133, pop: 100000 },
  { area: '大阪市天王寺区', pref: '大阪府', lat: 34.6532, lon: 135.5186, pop: 80000 },
  { area: '堺市', pref: '大阪府', lat: 34.5733, lon: 135.4832, pop: 830000 },
  // Other major cities
  { area: '名古屋市', pref: '愛知県', lat: 35.1815, lon: 136.9066, pop: 2330000 },
  { area: '札幌市', pref: '北海道', lat: 43.0618, lon: 141.3545, pop: 1970000 },
  { area: '福岡市', pref: '福岡県', lat: 33.5902, lon: 130.4017, pop: 1600000 },
  { area: '神戸市', pref: '兵庫県', lat: 34.6913, lon: 135.1830, pop: 1530000 },
  { area: '京都市', pref: '京都府', lat: 35.0116, lon: 135.7681, pop: 1460000 },
  { area: '広島市', pref: '広島県', lat: 34.3853, lon: 132.4553, pop: 1200000 },
  { area: '仙台市', pref: '宮城県', lat: 38.2682, lon: 140.8694, pop: 1090000 },
  { area: '千葉市', pref: '千葉県', lat: 35.6073, lon: 140.1063, pop: 980000 },
  { area: '北九州市', pref: '福岡県', lat: 33.8834, lon: 130.8752, pop: 940000 },
  { area: '新潟市', pref: '新潟県', lat: 37.9161, lon: 139.0364, pop: 790000 },
  { area: '浜松市', pref: '静岡県', lat: 34.7108, lon: 137.7261, pop: 790000 },
  { area: '熊本市', pref: '熊本県', lat: 32.8032, lon: 130.7079, pop: 740000 },
  { area: '岡山市', pref: '岡山県', lat: 34.6551, lon: 133.9195, pop: 720000 },
  { area: '那覇市', pref: '沖縄県', lat: 26.3344, lon: 127.6809, pop: 320000 },
];

const LISTING_TITLES = {
  furniture: ['ソファ 美品', 'テーブルセット', '本棚 引き取り希望', 'ベッドフレーム IKEA', '食器棚 無料'],
  electronics: ['iPhone 13 中古', 'PS5 本体', 'MacBook Air 2023', 'Switch 本体+ソフト', 'iPad mini'],
  clothing: ['ブランドバッグ', 'スニーカー Nike 27cm', 'コート 冬物', 'ヴィンテージ デニム', 'ドレス フォーマル'],
  vehicles: ['自転車 ママチャリ', 'ロードバイク GIANT', '原付 スクーター', '軽自動車 車検付き', '電動自転車'],
  free_items: ['引っ越し処分 まとめて', '家電セット 0円', '子供服まとめ', '段ボール大量', '植木鉢セット'],
  barter: ['米と野菜交換', '英会話⇔日本語', 'ギター⇔ベース', '手作りケーキ交換', 'DIY手伝い交換'],
};

function seededRandom(seed) {
  let x = Math.sin(seed * 9301 + 49297) * 233280;
  return x - Math.floor(x);
}

function generateSeedData() {
  const features = [];
  const now = new Date();
  let idx = 0;
  const totalPop = JAPAN_AREAS.reduce((s, a) => s + a.pop, 0);

  for (const area of JAPAN_AREAS) {
    const count = Math.max(2, Math.round((area.pop / totalPop) * 250));
    for (let j = 0; j < count && features.length < 250; j++) {
      idx++;
      const r1 = seededRandom(idx * 3);
      const r2 = seededRandom(idx * 7);
      const r3 = seededRandom(idx * 11);
      const r4 = seededRandom(idx * 13);

      const lat = area.lat + (r1 - 0.5) * 0.03;
      const lon = area.lon + (r2 - 0.5) * 0.04;

      const platform = PLATFORMS[Math.floor(r3 * PLATFORMS.length)];
      const category = LISTING_CATEGORIES[Math.floor(r4 * LISTING_CATEGORIES.length)];
      const titles = LISTING_TITLES[category] || LISTING_TITLES.furniture;
      const title = titles[Math.floor(seededRandom(idx * 17) * titles.length)];

      const daysAgo = Math.floor(seededRandom(idx * 19) * 30);
      const postDate = new Date(now - daysAgo * 86400000);

      const price = category === 'free_items' ? 0 :
        category === 'barter' ? 0 :
        Math.floor(seededRandom(idx * 23) * 50000) + 100;

      features.push({
        type: 'Feature',
        geometry: { type: 'Point', coordinates: [lon, lat] },
        properties: {
          id: `CL_${String(idx).padStart(5, '0')}`,
          platform,
          title,
          category,
          price,
          price_display: price === 0 ? '無料/交換' : `¥${price.toLocaleString()}`,
          area: area.area,
          prefecture: area.pref,
          condition: ['new', 'like_new', 'good', 'fair', 'poor'][Math.floor(seededRandom(idx * 29) * 5)],
          has_image: seededRandom(idx * 31) > 0.2,
          views: Math.floor(seededRandom(idx * 37) * 500),
          favorites: Math.floor(seededRandom(idx * 41) * 30),
          is_negotiable: seededRandom(idx * 43) > 0.5,
          timestamp: postDate.toISOString(),
          source: 'classifieds',
        },
      });
    }
  }
  return features.slice(0, 250);
}

import { fetchText } from './_liveHelpers.js';

const AREA_BY_PREF = Object.fromEntries(JAPAN_AREAS.map((a) => [a.pref, a]));

function findArea(text) {
  if (!text) return null;
  for (const a of JAPAN_AREAS) {
    if (text.includes(a.area)) return a;
  }
  for (const a of JAPAN_AREAS) {
    if (text.includes(a.pref)) return a;
  }
  return null;
}

/**
 * Iterate Jmty (ジモティー) listings nationwide. Jmty has per-prefecture
 * URLs like https://jmty.jp/{pref-slug}/sale. We hit ~12 large areas and
 * paginate up to 5 pages each.
 */
async function tryJmtyScrape() {
  const slugs = ['tokyo', 'kanagawa', 'osaka', 'aichi', 'hokkaido', 'fukuoka',
                 'hyogo', 'chiba', 'saitama', 'kyoto', 'hiroshima', 'miyagi'];
  const features = [];
  let idx = 0;
  for (const slug of slugs) {
    for (let page = 1; page <= 5; page++) {
      const url = `https://jmty.jp/${slug}/sale?page=${page}`;
      const html = await fetchText(url, { timeoutMs: 12_000, retries: 1 });
      if (!html) break;
      // Each item: link + title + city/area span
      const items = html.split(/p-articles-list-item/).slice(1);
      if (items.length === 0) break;
      let pageMatched = 0;
      for (const block of items) {
        const titleMatch = block.match(/p-item-title[^>]*>([^<]+)</);
        const priceMatch = block.match(/p-item-most-important[^>]*>([^<]+)</);
        const areaMatch = block.match(/p-item-area[^>]*>([^<]+)</) || block.match(/データから[^>]*>([^<]+)</);
        const linkMatch = block.match(/href="([^"]+)"/);
        const title = titleMatch?.[1]?.trim();
        if (!title) continue;
        const areaTxt = areaMatch?.[1]?.trim() || '';
        const area = findArea(areaTxt) || findArea(slug) || AREA_BY_PREF[areaTxt];
        if (!area) continue;
        const jLat = (Math.random() - 0.5) * 0.03;
        const jLon = (Math.random() - 0.5) * 0.04;
        idx++;
        features.push({
          type: 'Feature',
          geometry: { type: 'Point', coordinates: [area.lon + jLon, area.lat + jLat] },
          properties: {
            id: `JMTY_${idx}`,
            platform: 'jmty',
            title,
            price_display: priceMatch?.[1]?.trim() || null,
            area: area.area,
            prefecture: area.pref,
            url: linkMatch?.[1] ? new URL(linkMatch[1], 'https://jmty.jp').href : null,
            source: 'jmty_live',
          },
        });
        pageMatched++;
      }
      if (pageMatched === 0) break;
    }
  }
  return features.length > 0 ? features : null;
}

export default async function collectClassifieds() {
  const live = await tryJmtyScrape();
  const features = live && live.length > 0 ? live : generateSeedData();
  return {
    type: 'FeatureCollection',
    features,
    _meta: {
      source: live ? 'jmty_live' : 'classifieds_seed',
      fetchedAt: new Date().toISOString(),
      recordCount: features.length,
      live: !!live,
      description: 'Japanese classifieds from Jmty (ジモティー) — live nationwide scrape, geocoded by area',
    },
  };
}

File: native/collectors/marketplace/sources/classifieds.c
/* collectors/marketplace/sources/classifieds.c
 * Port of server/src/collectors/classifieds.js (fetchText + regex HTML scrape).
 * Jmty (ジモティー) nationwide: 12 prefecture slugs x up to 5 pages of
 * https://jmty.jp/{slug}/sale?page={N}. Split body on "p-articles-list-item",
 * extract title / price / area / href per block, geocode by JAPAN_AREAS.
 * generateSeedData() + _meta NOT ported (live-only; 0 rows when upstream down).
 * Feature has no `id` member → featureUid = sha1(JSON.stringify{g,p}); the JS
 * applies random lat/lon jitter so upstream uids are non-deterministic too. */
#include "../../../source.h"
#include "../../../lib/feedlib.h"
#include "../../../lib/geojson.h"
#include "../../../third_party/cJSON.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct { const char *area, *pref; double lat, lon; } jp_area;

/* JAPAN_AREAS — verbatim, same order as the JS array. */
static const jp_area AREAS[] = {
  { "世田谷区", "東京都", 35.6461, 139.6531 },
  { "練馬区",   "東京都", 35.7356, 139.6518 },
  { "大田区",   "東京都", 35.5613, 139.7161 },
  { "足立区",   "東京都", 35.7752, 139.8046 },
  { "江戸川区", "東京都", 35.6928, 139.8682 },
  { "杉並区",   "東京都", 35.6994, 139.6365 },
  { "板橋区",   "東京都", 35.7512, 139.7090 },
  { "新宿区",   "東京都", 35.6938, 139.7036 },
  { "渋谷区",   "東京都", 35.6640, 139.6982 },
  { "中野区",   "東京都", 35.7073, 139.6638 },
  { "豊島区",   "東京都", 35.7263, 139.7171 },
  { "目黒区",   "東京都", 35.6338, 139.6980 },
  { "横浜市",   "神奈川県", 35.4437, 139.6380 },
  { "川崎市",   "神奈川県", 35.5309, 139.7030 },
  { "相模原市", "神奈川県", 35.5710, 139.3731 },
  { "大阪市北区",   "大阪府", 34.7055, 135.4983 },
  { "大阪市中央区", "大阪府", 34.6813, 135.5133 },
  { "大阪市天王寺区","大阪府", 34.6532, 135.5186 },
  { "堺市",     "大阪府", 34.5733, 135.4832 },
  { "名古屋市", "愛知県", 35.1815, 136.9066 },
  { "札幌市",   "北海道", 43.0618, 141.3545 },
  { "福岡市",   "福岡県", 33.5902, 130.4017 },
  { "神戸市",   "兵庫県", 34.6913, 135.1830 },
  { "京都市",   "京都府", 35.0116, 135.7681 },
  { "広島市",   "広島県", 34.3853, 132.4553 },
  { "仙台市",   "宮城県", 38.2682, 140.8694 },
  { "千葉市",   "千葉県", 35.6073, 140.1063 },
  { "北九州市", "福岡県", 33.8834, 130.8752 },
  { "新潟市",   "新潟県", 37.9161, 139.0364 },
  { "浜松市",   "静岡県", 34.7108, 137.7261 },
  { "熊本市",   "熊本県", 32.8032, 130.7079 },
  { "岡山市",   "岡山県", 34.6551, 133.9195 },
  { "那覇市",   "沖縄県", 26.3344, 127.6809 },
};
#define NAREA ((int)(sizeof(AREAS) / sizeof(AREAS[0])))

/* JS findArea(text): first area whose .area substring matches, else first
 * whose .pref substring matches, else null. */
static const jp_area *find_area(const char *text) {
  if (!text || !text[0]) return NULL;
  for (int i = 0; i < NAREA; i++)
    if (strstr(text, AREAS[i].area)) return &AREAS[i];
  for (int i = 0; i < NAREA; i++)
    if (strstr(text, AREAS[i].pref)) return &AREAS[i];
  return NULL;
}

/* JS AREA_BY_PREF = Object.fromEntries(map([pref,area])): on duplicate pref
 * the LAST array entry wins. */
static const jp_area *area_by_pref(const char *text) {
  if (!text || !text[0]) return NULL;
  const jp_area *hit = NULL;
  for (int i = 0; i < NAREA; i++)
    if (strcmp(AREAS[i].pref, text) == 0) hit = &AREAS[i];
  return hit;
}

/* Mirror block.match(/<needle>[^>]*>([^<]+)</): find needle, then the next
 * '>' (skipping a [^>]* run — no '>' inside), capture up to next '<', trim.
 * Returns malloc'd captured text or NULL. */
static char *cap_after(const char *block, const char *needle) {
  const char *h = strstr(block, needle);
  if (!h) return NULL;
  const char *p = h + strlen(needle);
  /* [^>]* then '>' */
  while (*p && *p != '>') p++;
  if (*p != '>') return NULL;
  p++;
  /* ([^<]+) */
  const char *s = p;
  while (*p && *p != '<') p++;
  if (p == s) return NULL;
  size_t len = (size_t)(p - s);
  /* trim() */
  const char *a = s;
  const char *b = s + len;
  while (a < b && (unsigned char)*a <= ' ') a++;
  while (b > a && (unsigned char)b[-1] <= ' ') b--;
  if (b == a) return NULL;
  size_t n = (size_t)(b - a);
  char *out = malloc(n + 1);
  memcpy(out, a, n);
  out[n] = 0;
  return out;
}

/* Mirror block.match(/href="([^"]+)"/): first href value in the block. */
static char *cap_href(const char *block) {
  const char *h = strstr(block, "href=\"");
  if (!h) return NULL;
  const char *s = h + 6;
  const char *p = strchr(s, '"');
  if (!p || p == s) return NULL;
  size_t n = (size_t)(p - s);
  char *out = malloc(n + 1);
  memcpy(out, s, n);
  out[n] = 0;
  return out;
}

/* new URL(href, 'https://jmty.jp').href — resolve relative refs. Enough of
 * URL resolution for the shapes Jmty emits (absolute, scheme-relative,
 * root-relative, or path-relative). */
static char *resolve_url(const char *href) {
  if (!href || !href[0]) return NULL;
  if (strncmp(href, "http://", 7) == 0 || strncmp(href, "https://", 8) == 0)
    return strdup(href);
  char *out;
  if (strncmp(href, "//", 2) == 0) {
    size_t n = strlen(href) + 8;
    out = malloc(n);
    snprintf(out, n, "https:%s", href);
  } else if (href[0] == '/') {
    size_t n = strlen(href) + 24;
    out = malloc(n);
    snprintf(out, n, "https://jmty.jp%s", href);
  } else {
    size_t n = strlen(href) + 24;
    out = malloc(n);
    snprintf(out, n, "https://jmty.jp/%s", href);
  }
  return out;
}

static const char *SLUGS[] = {
  "tokyo", "kanagawa", "osaka", "aichi", "hokkaido", "fukuoka",
  "hyogo", "chiba", "saitama", "kyoto", "hiroshima", "miyagi"
};
#define NSLUG ((int)(sizeof(SLUGS) / sizeof(SLUGS[0])))

static int run(const source_ctx *ctx, intel_sink *sink) {
  cJSON *features = cJSON_CreateArray();
  int idx = 0;

  for (int s = 0; s < NSLUG; s++) {
    const char *slug = SLUGS[s];
    for (int page = 1; page <= 5; page++) {
      char url[160];
      snprintf(url, sizeof url,
               "https://jmty.jp/%s/sale?page=%d", slug, page);
      char *html = feed_get_text(ctx->http, url, 12000);
      if (!html) break;                          /* if(!html) break */

      /* html.split(/p-articles-list-item/).slice(1) — iterate blocks after
       * each occurrence; a block ends at the next occurrence (or EOF). */
      const char *MARK = "p-articles-list-item";
      size_t ml = strlen(MARK);
      const char *first = strstr(html, MARK);
      if (!first) { free(html); break; }         /* items.length===0 → break */

      int page_matched = 0;
      const char *cur = first + ml;
      while (cur) {
        const char *next = strstr(cur, MARK);
        size_t blen = next ? (size_t)(next - cur) : strlen(cur);
        char *block = malloc(blen + 1);
        memcpy(block, cur, blen);
        block[blen] = 0;

        char *title  = cap_after(block, "p-item-title");
        char *price  = cap_after(block, "p-item-most-important");
        char *areatx = cap_after(block, "p-item-area");
        if (!areatx) areatx = cap_after(block, "データから");
        char *href   = cap_href(block);

        if (title && title[0]) {
          const char *areatxt = areatx ? areatx : "";
          const jp_area *area = find_area(areatxt);
          if (!area) area = find_area(slug);
          if (!area) area = area_by_pref(areatxt);
          if (area) {
            double jLat = ((double)rand() / RAND_MAX - 0.5) * 0.03;
            double jLon = ((double)rand() / RAND_MAX - 0.5) * 0.04;
            idx++;

            cJSON *f = cJSON_CreateObject();
            cJSON_AddStringToObject(f, "type", "Feature");
            cJSON *g = cJSON_CreateObject();
            cJSON_AddStringToObject(g, "type", "Point");
            cJSON *co = cJSON_CreateArray();
            cJSON_AddItemToArray(co, cJSON_CreateNumber(area->lon + jLon));
            cJSON_AddItemToArray(co, cJSON_CreateNumber(area->lat + jLat));
            cJSON_AddItemToObject(g, "coordinates", co);
            cJSON_AddItemToObject(f, "geometry", g);

            cJSON *p = cJSON_CreateObject();   /* EXACT JS key order */
            char id[32];
            snprintf(id, sizeof id, "JMTY_%d", idx);
            cJSON_AddStringToObject(p, "id", id);
            cJSON_AddStringToObject(p, "platform", "jmty");
            cJSON_AddStringToObject(p, "title", title);
            if (price && price[0])
              cJSON_AddStringToObject(p, "price_display", price);
            else
              cJSON_AddItemToObject(p, "price_display", cJSON_CreateNull());
            cJSON_AddStringToObject(p, "area", area->area);
            cJSON_AddStringToObject(p, "prefecture", area->pref);
            char *rurl = (href && href[0]) ? resolve_url(href) : NULL;
            if (rurl)
              cJSON_AddStringToObject(p, "url", rurl);
            else
              cJSON_AddItemToObject(p, "url", cJSON_CreateNull());
            free(rurl);
            cJSON_AddStringToObject(p, "source", "jmty_live");
            cJSON_AddItemToObject(f, "properties", p);
            cJSON_AddItemToArray(features, f);
            page_matched++;
          }
        }

        free(title); free(price); free(areatx); free(href);
        free(block);
        cur = next ? next + ml : NULL;
      }
      free(html);
      if (page_matched == 0) break;              /* if(pageMatched===0) break */
    }
  }

  int n = geojson_emit_features(sink, ctx->source_id, features);
  cJSON_Delete(features);
  fprintf(stderr, "[classifieds] emitted %d\n", n);
  return n >= 0 ? 0 : -1;
}

static const source_def classifieds_def = {
  .id = "classifieds", .collector = "marketplace",
  .name = "Jmty / Mercari / Yahoo Auctions",
  .name_ja = "ジモティー・メルカリ・ヤフオク",
   .update_interval_sec = 1800, .run = run,
};
REGISTER_SOURCE(classifieds_def)
